### PR TITLE
Safer recovery

### PR DESCRIPTION
**Pull request by kareemk**

### CODE REVIEW

File: gemfiles/mongoid40.gemfile [hunk 1]
```diff
@@ -17,3 +17,5 @@ end
 group :development do
   gem 'pry'
 end
+
+gem 'robust-redis-lock', :path => '../robust-redis-lock'
```
Comment (kareemk) [inline gemfiles/mongoid40.gemfile:21]: Need to push out a gem. Doing that now and then I can delete this.

File: lib/promiscuous/amqp/bunny.rb [hunk 2]
```diff
@@ -82,14 +82,30 @@ def publish(options={})
     Promiscuous.debug "[publish] #{options[:exchange]}/#{options[:key]} #{options[:payload]}"
 
     @connection_lock.synchronize do
-      tag = @channel.next_publish_seq_no if options[:on_confirm]
-      @callback_mapping[tag] = options[:on_confirm] if options[:on_confirm]
+      if options[:async]
```
Comment (kareemk) [inline lib/promiscuous/amqp/bunny.rb:85]: fyi - i made AMQP.publish sync by default

File: lib/promiscuous/config.rb [hunk 3]
```diff
@@ -63,7 +63,7 @@ def self._configure(&block)
     self.publisher_lock_expiration ||= 5.seconds
     self.publisher_lock_timeout ||= 2.seconds
     self.recovery_interval    ||= 5.seconds
-    self.logger               ||= defined?(Rails) ? Rails.logger : Logger.new(STDERR).tap { |l| l.level = Logger::WARN }
+    self.logger               ||= Logger.new(STDERR).tap { |l| l.level = ENV["LOGGER_LEVEL"].to_i || Logger::WARN }
```
Comment (kareemk) [inline lib/promiscuous/config.rb:66]: revert

Comment (michaeldauria) [inline lib/promiscuous/config.rb:66]: should probably put this back

File: lib/promiscuous/amqp/bunny.rb (hunk 2 above)
Comment (michaeldauria) [inline lib/promiscuous/amqp/bunny.rb:85]: you mean async?

File: lib/promiscuous/publisher/operation/base.rb [hunk 4]
```diff
@@ -68,38 +68,19 @@ def lock_operations_and_queue_recovered_payloads
...
       rescue Redis::Lock::Recovered
-        recover_for_lock(lock)
-        begin
-          lock.extend
-        rescue
-          unlock_all_locks
-          raise Promiscuous::Error::LockUnavailable.new(lock.key)
-        end
+        Promiscuous::Publisher::Operation::Recovery.new(:lock => lock).recover!
+        retry
```
Comment (michaeldauria) [inline lib/promiscuous/publisher/operation/base.rb:76]: do you want to limit the # of retries?

File: lib/promiscuous/publisher/operation/base.rb [hunk 5]
```diff
@@ -132,26 +113,19 @@ def payloads
...
-    raise_error = options[:raise_error].present? ? options[:raise_error] : false
+    async       = !!options[:async]
 
     payloads.each do |payload|
-      begin
-        Promiscuous::AMQP.publish(:exchange => exchange.to_s,
-                                  :key => routing.to_s,
-                                  :payload => payload,
-                                  :on_confirm => method(:unlock_all_locks))
-      rescue Exception => e
```
Comment (michaeldauria) [inline lib/promiscuous/publisher/operation/base.rb:148]: you removed this exception handler, but in some cases nothing is catching it when `publish_payloads` is called ex: the next file `def execute_instrumented`.

File: lib/promiscuous/amqp/bunny.rb (hunk 2 above)
Comment (kareemk) [inline lib/promiscuous/amqp/bunny.rb:85]: no, sync

File: lib/promiscuous/publisher/operation/base.rb (hunk 4 above)
Comment (kareemk) [inline lib/promiscuous/publisher/operation/base.rb:76]: no. it will only happen once.

File: lib/promiscuous/publisher/operation/base.rb (hunk 5 above)
Comment (kareemk) [inline lib/promiscuous/publisher/operation/base.rb:148]: `AMQP.publish` handles this. It doesn't raise if its async. Otherwise it raises.

**Pull request closed without merging**